xformers/components/attention/csrc/cuda/attention.cu

```diff
@@ -577,14 +579,611 @@ at::Tensor attention(
+    at::PackedTensorAccessor<scalar_t, 3> value,
+    at::PackedTensorAccessor<scalar_t, 2> tmp_sum_i,
+    at::PackedTensorAccessor<scalar_t, 2> logsumexp_normalizer) {
+  int64_t K = query.size(2);
+  int64_t B = query.size(0);
+  int64_t M = query.size(1);
+  int64_t N = key.size(1);
+
+  constexpr int kVecSize = sizeof(vec_t) / sizeof(scalar_t);
+
+  int64_t batch_idx = blockIdx.z;
+  int64_t query_idx =
+      blockIdx.x * blockDim.x * kBlockSizeQ + threadIdx.x * kBlockSizeQ;
+  int64_t l = blockIdx.y * blockDim.y * kBlockSizeK + threadIdx.y * kBlockSizeK;
+
+  __shared__ scalar_t fact[TILE_SIZEQ][TILE_SIZEK + 1];
+
+#pragma unroll
+  for (int k_item_idx = 0; k_item_idx < kBlockSizeK; k_item_idx++) {
+#pragma unroll
+    for (int q_item_idx = 0; q_item_idx < kBlockSizeQ; q_item_idx++) {
+      fact[kBlockSizeQ * threadIdx.x + q_item_idx]
+          [kBlockSizeK * threadIdx.y + k_item_idx] = 0;
+    }
+  }
+
+  scalar_t normalizer[kBlockSizeQ];
+  scalar_t tmp_sum[kBlockSizeQ] = {0};
+
+  vec_t *qb[kBlockSizeQ], *kb[kBlockSizeK], *vb[kBlockSizeK], *gb[kBlockSizeQ],
+      *gbb[TILE_SIZEQ];
+  scalar_t maskQ[kBlockSizeQ], maskK[kBlockSizeK];
+
+  for (int k_item_idx = 0; k_item_idx < kBlockSizeK; k_item_idx++) {
+    int64_t index = l + k_item_idx;
+    maskK[k_item_idx] = index >= N ? scalar_t(0) : scalar_t(1);
+    if (check_bounds)
+      index = min(index, N - 1);
+    kb[k_item_idx] = reinterpret_cast<vec_t*>(key[batch_idx][index].data());
+    vb[k_item_idx] = reinterpret_cast<vec_t*>(value[batch_idx][index].data());
+  }
+
+  for (int q_item_idx = 0; q_item_idx < kBlockSizeQ; q_item_idx++) {
+    int64_t index = query_idx + q_item_idx;
+    maskQ[q_item_idx] = index >= M ? scalar_t(0) : scalar_t(1);
+    if (check_bounds)
+      index = min(index, M - 1);
+    qb[q_item_idx] = reinterpret_cast<vec_t*>(query[batch_idx][index].data());
+    gb[q_item_idx] =
+        reinterpret_cast<vec_t*>(grad_out[batch_idx][index].data());
+  }
+
+  for (int64_t i = 0; i < TILE_SIZEQ; i++) {
+    int64_t index = query_idx + i - kBlockSizeQ * threadIdx.x;
+    if (check_bounds)
+      index = min(index, M - 1);
+    gbb[i] = reinterpret_cast<vec_t*>(grad_out[batch_idx][index].data());
+  }
+
+  for (int i = 0; i < kBlockSizeQ; i++) {
+    int64_t index = query_idx + i;
+    if (check_bounds)
+      index = min(index, M - 1);
+    normalizer[i] = logsumexp_normalizer[batch_idx][index];
+  }
+
+  scalar_t attn_v[kBlockSizeQ][kBlockSizeK] = {0};
+  scalar_t grad_attn_v[kBlockSizeQ][kBlockSizeK] = {0};
+  scalar_t scale = 1.0 / std::sqrt(scalar_t(K));
+
+  for (int64_t k = 0; k < K / kVecSize; k += 1) {
+#pragma unroll
+    for (int k_item_idx = 0; k_item_idx < kBlockSizeK; k_item_idx++) {
+      vec_t kk = __ldg(kb[k_item_idx] + k);
+      iMul(scale, &kk);
+      vec_t tt = __ldg(vb[k_item_idx] + k);
+#pragma unroll
+      for (int q_item_idx = 0; q_item_idx < kBlockSizeQ; q_item_idx++) {
+        sputnik::VectorCompute<vec_t>::Dot(
```

Comment:
(I guess that `computeDot` is the most interesting place for that, but I cannot comment there.) It could be interesting to pull in tensor cores for the .dot() for fp16, and for fp32 on newer hardware (A100 +), I don't know if there are existing primitives to preferably use there (from Cuda/cudnn/cublas, [like in this example](https://developer.nvidia.com/blog/programming-tensor-cores-cuda-9/) or from torch). 

Maybe that it's not needed actually, sorry for bringing these tensor cores up all the time but just trying to think ahead for the possible next steps (and possibly hidden caveats if you developed on a P100)

Reply:
TensorCores are going to be a very important next step indeed, and I'm already looking into how to use them. But before that I think it might be better to support the `K > 32` case first, wdyt?

Reply:
agree on bigger K first, at this point the perf impact is not super clear to me (for fp32 at least, on the 3080 which has tensor cores this was very competitive vs. pytorch and tensor cores) while the limitation of a small K is very well defined